src/python/turicreate/toolkits/sound_classifier/sound_classifier.py

```diff
@@ -806,35 +805,39 @@ def get_custom_model_spec():
             builder = NeuralNetworkBuilder([(input_name, Array(input_length,))],
                                            [(prob_name, Dictionary(String))],
                                            'classifier')
+            layer_counter = [0]
+            builder.set_input([input_name], [(input_length,)])
+
+            def next_layer_name():
+                layer_counter[0] += 1
+                return "layer_%d" % layer_counter[0]
 
-            input_name, output_name = input_name, 0
             for i, cur_layer in enumerate(self._custom_classifier.export_weights()):
                 W = cur_layer['weight']
                 nC, nB = W.shape
                 Wb = cur_layer['bias']
 
+                output_name = next_layer_name()
                 builder.add_inner_product(name="inner_product_"+str(i),
                                           W=W,
                                           b=Wb,
                                           input_channels=nB,
                                           output_channels=nC,
                                           has_bias=True,
-                                          input_name=str(input_name),
-                                          output_name='inner_product_'+str(output_name))
-
-                if cur_layer['act']:
-                    builder.add_activation("activation"+str(i), 'RELU', 'inner_product_'+str(output_name), str(output_name))
+                                          input_name=input_name,
+                                          output_name=output_name)
 
-                input_name = i
-                output_name = i + 1
-
-            last_output = builder.spec.neuralNetworkClassifier.layers[-1].output[0]
-            builder.add_softmax('softmax', last_output, self.target)
-
-            builder.set_class_labels(self.classes, predicted_feature_name = self.target)
-            builder.set_input([input_name], [(input_length,)])
-            builder.set_output([self.target], [(self.num_classes,)])
+                input_name = output_name
 
+                if cur_layer['act']:
+                    output_name = next_layer_name()
+                    builder.add_activation("activation"+str(i), 'RELU', input_name, output_name)
+                    input_name = output_name
+
+            builder.add_softmax('softmax', input_name, self.target)
```

Comment:
The bug was simply a difference in behavior between 3.0b1 and 3.1 where setting class labels before output vs after output overrides the class names.

Reply:
Is this something that worth a coremltools GitHub issue? Or do you think the coremltools behavior is better in 3.1? Overriding the class names sounds bad to me.

Reply:
No I mis-understood the behavior. The new API is actually cleaner.